File: backend/src/algorithms/tokenBucket.ts
//data is not lost

import type { Result } from "../types/index.js";


// export async function tokenBucketAlgorithm(
//   key: string,
//   limit: number, //max tokens  a bucket can hold
//   refillRatePerSecond: number,   
//   tokensRequested:number,   
// ): Promise<Result> {
//   try {
//     const redisKey = `token_bucket:{${key}}`;
//     const lastRefillKey = `${redisKey}:last_refill`;
//     const currentTime = Math.floor(Date.now() / 1000);
//     const statsKey = `stats:{${key}}`;          // per-user stats
// const globalStatsKey = `stats:{global}`;   // global stats


  
//     const lastRefill = await redis.get(lastRefillKey);
//     const lastRefillTime: number = lastRefill ? parseInt(lastRefill) : currentTime;

  
//     let currentCountStr = await redis.get(redisKey);
//     let currentCount: number = currentCountStr ? parseInt(currentCountStr) : limit;

//     const elapsed = currentTime - lastRefillTime;
//     const tokensToAdd = Math.floor(elapsed * refillRatePerSecond);
//     currentCount = Math.min(limit, currentCount + tokensToAdd);

  
//     if (currentCount < tokensRequested) {
//     let blocked=  await redis.hincrby(statsKey, 'blocked', 1);
//      let total= await redis.hincrby(statsKey, 'total', 1);
//       await redis.hincrby(globalStatsKey, 'blocked', 1);
//       await redis.hincrby(globalStatsKey, 'total', 1);
 
//       return {
//          allowed:false,
//         algorithmName: 'token_bucket',
//          tokensRemaining:currentCount,
   
//        // result stats per key
//          blockedRequests: blocked,
//          totalRequests:total,
//          allowedRequests:total - blocked,
//         retryAfterTime: Math.ceil((tokensRequested - currentCount) / refillRatePerSecond),
//       };
//     }

  
//     currentCount -= tokensRequested;
//     let allowed=await redis.hincrby(statsKey, 'allowed', 1);
//     let total= await redis.hincrby(statsKey, 'total', 1);
//     await redis.hincrby(globalStatsKey, 'allowed', 1);
//     await redis.hincrby(globalStatsKey, 'total', 1);
 
//     await redis.set(redisKey, currentCount);
//     await redis.set(lastRefillKey, currentTime.toString());

//     return {
//        allowed: true,
//         algorithmName: 'token_bucket',
//        tokensRemaining: currentCount,
     
//      // result stats per key
//        blockedRequests: total - allowed,
//        totalRequests: total,
//        allowedRequests: allowed,
//     };
//   } catch (err) {
//   console.error("Token Bucket Error:", err);

//   return {
//     allowed: true,
//      algorithmName: 'token_bucket',
//     tokensRemaining: limit,
//     blockedRequests: 0,
//     totalRequests: 0,
//     allowedRequests: 0,
//     };
//   }
// }

import { redis } from "../utils/redis.js";
import fs from "fs";
import path from "path";

let tokenBucketSHA: string | null = null;


async function loadTokenBucketScript() {
  if (!tokenBucketSHA) {
    const luaScript = fs.readFileSync(
      path.join(process.cwd(), "src/algo-lua/tokenBucket.lua"),
      "utf8"
    );

   tokenBucketSHA = await (redis as any).script("load", luaScript);

    console.log("Token Bucket Script SHA Loaded:", tokenBucketSHA);
  }
  return tokenBucketSHA;
}
export async function tokenBucketAlgorithm(
  key: string,
  limit: number,
  refillRatePerSecond: number,
  tokensRequested: number
) {
  await loadTokenBucketScript();

const redisKey = `token_bucket:{${key}}`;
const statsKey = `stats:{${key}}`;
const globalStatsKey = `stats:{${key}}:global`;

  const now = Math.floor(Date.now() / 1000);

  const raw:any = await redis.evalsha(
    tokenBucketSHA!,
    3,
    redisKey,
    statsKey,
    globalStatsKey,
    limit.toString(),
    refillRatePerSecond.toString(),
    now.toString(),
    tokensRequested.toString()
  );

  console.log("RAW REDIS RESULT:", raw);

  const allowed = raw[0] === 1;

  return {
    allowed,
    algorithmName: "token bucket",

    tokensRemaining: raw[1],
    retryAfterTime: raw[2],
    resetTime: raw[3],

    blockedRequests: raw[4],
    totalRequests: raw[5],
    allowedRequests: raw[6],
  };
}

File: backend/src/algo-lua/tokenBucket.lua
---@diagnostic disable: undefined-global

local redisKey = KEYS[1]
local statsKey = KEYS[2]
local globalStats = KEYS[3]

local limit = tonumber(ARGV[1]) or 0
local refillRatePerSec = tonumber(ARGV[2])
local currentTime = tonumber(ARGV[3])
local tokenRequested = tonumber(ARGV[4])

local lastRefillKey = redisKey .. ":last_refill"


local lastRefillTime = tonumber(redis.call("GET", lastRefillKey))
if not lastRefillTime then
    lastRefillTime = currentTime
end


local currentTokens = tonumber(redis.call("GET", redisKey)) or limit


local elapsedSec = (currentTime - lastRefillTime) / 1000
local tokensToAdd = math.floor(elapsedSec * refillRatePerSec)
currentTokens = math.min(limit, currentTokens + tokensToAdd)
if tokensToAdd > 0 then
  redis.call("SET", lastRefillKey, currentTime)
end

if tokenRequested > currentTokens then
    local retryAfterMs = math.ceil((tokenRequested - currentTokens) / refillRatePerSec * 1000)
    return {
        0,
        currentTokens,
        retryAfterMs,
        currentTime + retryAfterMs,
        blocked,
        total,
        total - blocked
    }
end



currentTokens = currentTokens - tokenRequested
redis.call("SET", redisKey, currentTokens)
redis.call("PEXPIRE", redisKey, math.ceil((limit / refillRatePerSec) * 1000))

local allowed = redis.call("HINCRBY", statsKey, "allowed", 1)
local total = redis.call("HINCRBY", statsKey, "total", 1)

redis.call("HINCRBY", globalStats, "allowed", 1)
redis.call("HINCRBY", globalStats, "total", 1)

return {
    1,                                      
    currentTokens,                         
    0,                                     
    currentTime,                            
    total - allowed,                    
    total,                                  
    allowed
      
}